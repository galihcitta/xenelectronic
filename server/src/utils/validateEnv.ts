import { port, cleanEnv } from 'envalid';

export const validateEnv = () => {
  cleanEnv(process.env, {
    PORT: port()
  });
};

// https://www.npmjs.com/package/envalid
