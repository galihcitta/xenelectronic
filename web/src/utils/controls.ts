import { Controls } from '../pages/Auth/types';

export const signInControls: Controls = {
  email: {
    label: 'Enter email',
    placeholder: '',
    type: 'email',
    id: 'signin-email',
    name: 'email',
    value: '',
    validationRules: {
      required: true,
      isEmail: true
    },
    valid: false,
    touched: false,
    errorMsg: ''
  },
  password: {
    label: 'Enter password',
    placeholder: '',
    type: 'password',
    id: 'signin-password',
    name: 'password',
    value: '',
    validationRules: {
      required: true
    },
    valid: false,
    touched: false,
    errorMsg: ''
  }
};

export const signUpControls: Controls = {
  username: {
    label: 'Enter username',
    placeholder: '',
    type: 'text',
    id: 'signup-username',
    name: 'username',
    value: '',
    validationRules: {
      required: true,
      minLength: 4,
      maxLength: 20
    },
    valid: false,
    touched: false,
    errorMsg: ''
  },
  email: {
    label: 'Enter email',
    placeholder: '',
    type: 'email',
    id: 'signup-email',
    name: 'email',
    value: '',
    validationRules: {
      required: true,
      isEmail: true
    },
    valid: false,
    touched: false,
    errorMsg: ''
  },
  password: {
    label: 'Enter password',
    placeholder: '',
    type: 'password',
    id: 'signup-password',
    name: 'password',
    value: '',
    validationRules: {
      required: true,
      isSecure: true,
      minLength: 8,
      maxLength: 20
    },
    valid: false,
    touched: false,
    errorMsg: ''
  }
};
