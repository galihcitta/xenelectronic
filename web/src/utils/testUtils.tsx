import React from 'react';
import { BrowserRouter, Router } from 'react-router-dom';
import { Provider } from 'react-redux';
import { store } from '../modules/store';
import { createMemoryHistory } from 'history';

const history = createMemoryHistory();

export const withRouter = <P extends object>(
  Component: React.ComponentType<P>
) => (props: P) => (
  <BrowserRouter>
    <Component {...props} />
  </BrowserRouter>
);

export const withHistory = <P extends object>(
  Component: React.ComponentType<P>
) => (props: P) => (
  <Router history={history}>
    <Component {...props} />
  </Router>
);

export const withProvider = <P extends object>(
  Component: React.ComponentType<P>
) => (props: P) => (
  <Provider store={store}>
    <BrowserRouter>
      <Component {...props} />
    </BrowserRouter>
  </Provider>
);
