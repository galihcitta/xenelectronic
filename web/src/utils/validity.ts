import { ValidationRules } from '../pages/Auth/types';

export const validate = (
  value: string,
  controlName: string,
  rules: ValidationRules
) => {
  let isValid = true;
  let errorMsg = '';

  if (rules.required) {
    isValid = value.trim() !== '' && isValid;
    if (!isValid && !errorMsg.length)
      errorMsg = `${controlName} field is required`;
  }

  if (rules.minLength) {
    isValid = value.length >= rules.minLength && isValid;
    if (!isValid && !errorMsg.length)
      errorMsg = `${controlName} must be at least ${rules.minLength} characters`;
  }

  if (rules.maxLength) {
    isValid = value.length <= rules.maxLength && isValid;
    if (!isValid && !errorMsg.length)
      errorMsg = `${controlName} must be no longer than ${rules.maxLength} characters`;
  }

  if (rules.isEmail) {
    const pattern = /[a-z0-9!#$%&'*+/=?^_`{|}~-]+(?:\.[a-z0-9!#$%&'*+/=?^_`{|}~-]+)*@(?:[a-z0-9](?:[a-z0-9-]*[a-z0-9])?\.)+[a-z0-9](?:[a-z0-9-]*[a-z0-9])?/;
    isValid = pattern.test(value) && isValid;
    if (!isValid && !errorMsg.length) errorMsg = `email is invalid`;
  }

  if (rules.isSecure) {
    // Minimum eight characters, at least one uppercase letter, one lowercase letter, one number and one special character
    const pattern = /^(?=.*[a-z])(?=.*[A-Z])(?=.*\d)(?=.*[@$!%*?&])[A-Za-z\d@$!%*?&]{8,}$/;
    isValid = pattern.test(value) && isValid;
    if (!isValid && !errorMsg.length) errorMsg = `password too weak`;
  }

  return {
    isValid,
    errorMsg: errorMsg.charAt(0).toUpperCase() + errorMsg.slice(1)
  };
};
