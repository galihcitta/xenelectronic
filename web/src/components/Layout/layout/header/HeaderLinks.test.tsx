import { render, fireEvent } from '@testing-library/react';
import { _HeaderLinks, HeaderLinksProps } from './HeaderLinks.component';
import { withRouter } from '../../../../utils/testUtils';

const HeaderLinksWithRouter = withRouter(_HeaderLinks);

describe('<HeaderLinks />', () => {
  const mockCheckSession = jest.fn();
  const mockFetchProductsByType = jest.fn();

  const props: HeaderLinksProps = {
    isAuth: false,
    cartItemsCount: 2,
    checkSession: mockCheckSession,
    fetchProductsByType: mockFetchProductsByType
  };

  const setup = (props: HeaderLinksProps) => {
    const utils = render(HeaderLinksWithRouter(props));
    return { ...utils };
  };

  it('renders correctly', () => {
    const { asFragment } = setup(props);
    expect(asFragment()).toMatchSnapshot();
  });

  it('calls `fetchProductsByType` action with `all` arg', () => {
    const { getByText } = setup(props);
    fireEvent.click(getByText('Products'));
    expect(mockFetchProductsByType).toBeCalledWith('all');
  });

  it('calls `fetchProductsByType` action with `clocks` arg', () => {
    const { getByText } = setup(props);
    fireEvent.click(getByText('Clocks'));
    expect(mockFetchProductsByType).toBeCalledWith('clocks');
  });

  describe('when `isAuth` prop is true', () => {
    beforeEach(() => {
      props.isAuth = true;
    });

    it('renders `SignOutBtn` element', () => {
      const { getByText } = setup(props);
      expect(getByText('Logout')).toBeInTheDocument();
    });

    it('calls `chekcSession` action', () => {
      const { getByText } = setup(props);
      fireEvent.click(getByText('Logout'));
      expect(mockCheckSession).toHaveBeenCalled();
    });
  });
});
