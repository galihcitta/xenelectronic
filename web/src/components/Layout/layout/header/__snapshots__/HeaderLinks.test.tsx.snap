// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HeaderLinks /> renders correctly 1`] = `
<DocumentFragment>
  <ul
    class="sc-bdVaJa hDTYr"
  >
    <li
      class="sc-bxivhb icElMN"
    >
      <a
        class="sc-htpNat gyqNkk"
        href="/shop"
      >
        Products
        <i
          class="fas fa-caret-down"
        />
      </a>
      <div
        class="sc-bwzfXH gjnxxD"
      >
        <ul
          class="sc-ifAKCX kqqTha"
        >
          <li
            class="sc-EHOje jTwYBT"
          >
            <a
              class="sc-bZQynM imRHDl"
              href="/shop"
            >
              Chairs
            </a>
          </li>
          <li
            class="sc-EHOje jTwYBT"
          >
            <a
              class="sc-bZQynM imRHDl"
              href="/shop"
            >
              Lamps
            </a>
          </li>
          <li
            class="sc-EHOje jTwYBT"
          >
            <a
              class="sc-bZQynM imRHDl"
              href="/shop"
            >
              Clocks
            </a>
          </li>
          <li
            class="sc-EHOje jTwYBT"
          >
            <a
              class="sc-bZQynM imRHDl"
              href="/shop"
            >
              Vases
            </a>
          </li>
          <li
            class="sc-EHOje jTwYBT"
          >
            <a
              class="sc-bZQynM imRHDl"
              href="/shop"
            >
              Benches
            </a>
          </li>
          <li
            class="sc-EHOje jTwYBT"
          >
            <a
              class="sc-bZQynM imRHDl"
              href="/shop"
            >
              Flowerpots
            </a>
          </li>
          <li
            class="sc-EHOje jTwYBT"
          >
            <a
              class="sc-bZQynM imRHDl"
              href="/shop"
            >
              Plants
            </a>
          </li>
        </ul>
      </div>
    </li>
    <li
      class="sc-bxivhb icElMN"
    >
      <a
        class="sc-htpNat gyqNkk"
        href="/account"
      >
        Account
      </a>
    </li>
    <li
      class="sc-bxivhb icElMN"
    >
      <a
        class="sc-htpNat gyqNkk"
        href="/cart"
      >
        Cart (2)
      </a>
    </li>
  </ul>
</DocumentFragment>
`;
